src/main/java/edu/kit/kastel/sdq/intelligrade/extensions/guis/table/AnnotationsTableModel.java

```diff
@@ -0,0 +1,226 @@
+            Map<String, List<Annotation>> map = annotationsForClassifier.stream()
+                    .collect(Collectors.groupingBy(
+                            annotation -> {
+                                List<String> classifiers = annotation.getClassifiers();
+                                if (classifiers.size() <= 1) {
+                                    return annotation.getUUID();
+                                } else {
+                                    // to simplify the grouping code, we merge the remaining classifiers
+                                    // into a single string
+                                    return String.join(" ", classifiers.subList(1, classifiers.size()));
+                                }
+                            },
+                            LinkedHashMap::new,
+                            Collectors.toList()));
+
+            for (List<Annotation> list : map.values()) {
+                result.add(createNode(list));
+            }
+        }
+
+        return result;
+    }
+
+    // Don't ask, I couldn't think of a better solution.
+    private static AnnotationsTreeNode tempRootNode = null;
+
+    private static AnnotationsTreeNode storeRootNode(AnnotationsTreeNode node) {
+        tempRootNode = node;
+        return node;
+    }
+
+    public AnnotationsTableModel() {
+        super(storeRootNode(new AnnotationsTreeNode.RootNode()), AnnotationsTreeNode.columns());
+        this.root = tempRootNode;
+    }
+
+    private void refreshNodes() {
+        this.root.removeAllChildren();
+
+        for (var child : groupAnnotations(this.annotations)) {
+            this.root.add(child);
+        }
+
+        this.sort(AnnotationsTreeTable.DEFAULT_NODE_COMPARATOR);
+    }
+
+    public void sort(Comparator<? super AnnotationsTreeNode> comparator) {
+        // The root node and all its children are sorted here.
+        // An alternative approach would be to sort the list of annotations
+        // and then rebuild the tree, but this would destroy the tree structure
+        // and invalidate all the TreePaths.
+
+        this.root.sort(comparator);
+        this.reload();
+    }
+
+    @Override
+    public int getColumnCount() {
+        return AnnotationsTreeNode.numberOfColumns();
+    }
+
+    @Override
+    public String getColumnName(int column) {
+        return AnnotationsTreeNode.columnName(column);
+    }
+
+    @Override
+    public Object getValueAt(Object node, int column) {
+        return ((AnnotationsTreeNode) node).getValueAt(column);
+    }
+
+    public void setAnnotations(List<Annotation> annotations) {
+        // the annotations can change in the following ways:
+        // - annotations are added
+        // - annotations are removed
+        // - annotations are changed (e.g. the custom message is changed)
+        ///  ^ for these the identity of the annotation is unchanged -> we don't have to change anything
```

Comment:
Aren't annotations also grouped by their custom message?

Reply:
Yes they are, but these comments are just an overview of what the annotations can be:
- new ones that are currently not in the table
- some of the existing ones are missing from the current table
- some annotations have been updated (but because the object is modified and not a copy, this is not a problem)

Reply:
This is the code that defines which annotations are grouped:

https://github.com/kit-sdq/intelligrade/blob/35f2b729ae6dd508eec996046363ea5b710a630f/src/main/java/edu/kit/kastel/sdq/intelligrade/extensions/guis/table/AnnotationsTableModel.java#L56-L101

Maybe this explanation helps: The mentioned code is what can change between the currently present annotations (in the tree) and the new set of annotations, and this is:
- an annotation got added
- an annotation got removed
- an annotation got changed
or I might have misunderstood you